File: benchmark.cc
#include <boost/functional/hash.hpp>
#include <iostream>
#include <ctime>
#include <unordered_map>

constexpr int N_INTS = 2;
constexpr int N_KEYS = 1e7;

int main() {
  std::cout << "Start C++ STL benchmark:\n";

  typedef std::array<int, N_INTS> KeyType;
  std::unordered_map<std::array<int, N_INTS>, double, boost::hash<KeyType>> h;

  h.reserve(N_KEYS * 2);
  const double t0 = std::clock();

  KeyType key;
  for (int i = 1; i <= N_KEYS; i++) {
    for (int j = 1; j <= N_INTS; j++) {
      key[j - 1] = i + j;
    }
    h[key] = i * 0.5;
  }
  const double t1 = std::clock();

  double val;
  for (int i = 1; i <= N_KEYS; i++) {
    for (int j = 1; j <= N_INTS; j++) {
      key[j - 1] = i + j;
    }
    val = h[key];
  }
  const double t2 = std::clock();

  h.clear();
  const double t3 = std::clock();

  std::cout << "Time to assemble / get / clear:"
      << "       " << (t1 - t0) / CLOCKS_PER_SEC
      << "       " << (t2 - t1) / CLOCKS_PER_SEC
      << "       " << (t3 - t2) / CLOCKS_PER_SEC
      << "\n";

  return 0;
}

File: benchmark.f90
#define KEY_ARRAY_SIZE 2

#define FHASH_NAME int2real
#define KEY_TYPE integer, dimension(KEY_ARRAY_SIZE)
#define VALUE_TYPE real(real64)
#define VALUE_USE use, intrinsic :: iso_fortran_env, only: real64
#include "fhash.f90"

program test_benchmark
  implicit none
  
  call benchmark(n_ints=KEY_ARRAY_SIZE, n_keys=10**7)
  
contains
  subroutine benchmark(n_ints, n_keys)
    use int2real_mod
    use iso_fortran_env, only: real64

    integer, intent(in) :: n_ints, n_keys

    type(int2real_t) :: h
    integer :: key(n_ints)
    integer :: i, j
    real :: t0, t1, t2, t3
    real(real64), pointer :: val

    write(*,'(a)') "Start fhash benchmark:"

    write(*,'("n_ints: ", I0, ", n_keys: ", I0)') n_ints, n_keys

    call h%reserve(n_keys * 2)
    call cpu_time(t0)

    do i = 1, n_keys
      do j = 1, n_ints
        key(j) = i + j
      enddo
      call h%set(key, i * 0.5d0)
    enddo
    call cpu_time(t1)

    do i = 1, n_keys
      do j = 1, n_ints
        key(j) = i + j
      enddo
      val => h%get_ptr(key) ! , autoval=3.0_real64)
    enddo
    call cpu_time(t2)

    call h%clear()
    call cpu_time(t3)

    write(*,'(a,3(g15.3))') "Time to assemble/ get / clear: ", t1 - t0, t2 - t1, t3 - t2
  end subroutine
end program
